gobblin-service/src/test/java/org/apache/gobblin/service/modules/orchestration/MysqlDagStateStoreWithDagNodesTest.java

```diff
@@ -137,4 +152,56 @@ public void testAddGetAndDeleteDag() throws Exception{
     Assert.assertNull(this.dagStateStore.getDag(dagId1));
     Assert.assertNull(this.dagStateStore.getDag(dagId2));
   }
+
+  @Test
+  public void testMarkDagAsFailed() throws Exception {
+    // Set up initial conditions
+    Dag<JobExecutionPlan> dag = DagTestUtils.buildDag("test_dag", 789L);
+    DagManager.DagId dagId = DagManagerUtils.generateDagId(dag);
+
+    this.dagStateStore.writeCheckpoint(dag);
+
+    // Fetch all initial states into a list
+    List<Boolean> initialStates = fetchDagNodeStates(dagId.toString());
+
+    // Check Initial State
+    for (Boolean state : initialStates) {
+      Assert.assertFalse(state);
+    }
+    // Set the DAG as failed
+    dag.setFailedDag(true);
+    this.dagStateStore.writeCheckpoint(dag);
+
+    // Fetch all states after marking the DAG as failed
+    List<Boolean> failedStates = fetchDagNodeStates(dagId.toString());
+
+    // Check if all states are now true (indicating failure)
+    for (Boolean state : failedStates) {
+      Assert.assertTrue(state);
+    }
+    dagStateStore.cleanUp(dagId);
+    Assert.assertNull(this.dagStateStore.getDag(dagId));
+  }
+
+  private List<Boolean> fetchDagNodeStates(String dagId) throws IOException {
+    List<Boolean> states = new ArrayList<>();
+
+    dbStatementExecutor.withPreparedStatement(String.format(GET_DAG_NODES_STATEMENT, tableName), getStatement -> {
```

Comment:
is behind-the-scenes DB access the only way to validate behavior here?  is there no way to access from the "official" DagStateStore, then mark failed and finally re-access from the DSS to verify all nodes have changed?